File: backend/app/dto/tailoring.py
"""Typed data transfer objects for resume tailoring."""

from __future__ import annotations

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, field_validator

from app.enums.tailoring import TailoringStatus


class CoverLetterDTO(BaseModel):
    """Structured cover letter content generated for a tailoring session."""

    model_config = ConfigDict(frozen=True, extra="forbid")

    title: str = Field(min_length=1, max_length=255)
    greeting: str = Field(min_length=1)
    introduction: str = Field(min_length=1)
    body: str = Field(min_length=1)
    closing: str = Field(min_length=1)

    @field_validator("title", "greeting", "introduction", "body", "closing")
    @classmethod
    def _strip_text(cls, value: str) -> str:
        cleaned = value.strip()
        if not cleaned:
            raise ValueError("Text fields must not be blank")
        return cleaned


class ResumeVersionDTO(BaseModel):
    """Structured tailored resume version payload."""

    model_config = ConfigDict(frozen=True, extra="forbid")

    id: UUID | None = None
    resume_id: UUID | None = None
    tailoring_session_id: UUID | None = None
    professional_summary: str = Field(min_length=1)
    experience_json: list[dict[str, object]] = Field(default_factory=list)
    skills_json: list[dict[str, object]] = Field(default_factory=list)
    ats_score: int = Field(ge=0, le=100)
    recommendations_json: list[dict[str, object]] = Field(default_factory=list)
    created_at: datetime | None = None
    updated_at: datetime | None = None

    @field_validator("professional_summary")
    @classmethod
    def _strip_summary(cls, value: str) -> str:
        cleaned = value.strip()
        if not cleaned:
            raise ValueError("professional_summary must not be blank")
        return cleaned


class TailoringSessionDTO(BaseModel):
    """Tailoring session metadata and lifecycle state."""

    model_config = ConfigDict(frozen=True, extra="forbid")

    id: UUID
    resume_id: UUID
    job_description_id: UUID
    status: TailoringStatus
    created_at: datetime
    updated_at: datetime


class ResumeTailoringDTO(BaseModel):
    """Final combined output returned from resume tailoring service."""

    model_config = ConfigDict(frozen=True, extra="forbid")

    session: TailoringSessionDTO | None = None
    resume_version: ResumeVersionDTO
    cover_letter: CoverLetterDTO

File: frontend/src/types/tailoring.ts
// Mirrors backend/app/schemas/resume_tailoring.py and backend/app/enums/tailoring.py.
// NOTE: creating a session requires an existing job_description_id — see README.

export type TailoringStatus = "pending" | "processing" | "completed" | "failed";

export interface TailoringResumeVersionResponse {
  id: string;
  resume_id: string;
  tailoring_session_id: string;
  professional_summary: string;
  experience_json: Array<Record<string, unknown>>;
  skills_json: Array<Record<string, unknown>>;
  ats_score: number;
  recommendations_json: Array<Record<string, unknown>>;
  created_at: string;
  updated_at: string;
}

export interface CoverLetterResponse {
  id: string | null;
  tailoring_session_id: string | null;
  title: string;
  greeting: string;
  introduction: string;
  body: string;
  closing: string;
  created_at: string | null;
  updated_at: string | null;
}

export interface TailoringSessionResponse {
  id: string;
  resume_id: string;
  job_description_id: string;
  status: TailoringStatus;
  created_at: string;
  updated_at: string;
}

export interface TailoringSummaryResponse {
  session: TailoringSessionResponse;
  resume_version: TailoringResumeVersionResponse;
  cover_letter: CoverLetterResponse;
}

export type ExportFormat = "md" | "docx" | "pdf";
